Snapshot fixes

tiagolobocastro opened the pull request:
fix(init): ensure img-file has the correct size

    Fixes idempotency of volume creation in case file was created but
    not truncated to the size.

---

    test(smoke/snapshot): add snapshot deletion of unstaged volume

    Adds a test case where a volume is unstaged when its snapshot is deleted.

---

    fix(snapshot): idempotent snapshot delete

    On a retry the subvolume may already have been deleted.

---

    fix(snapshots): don't unmount volume on snap deletion

tiagolobocastro committed:
fix(snapshots): don't unmount volume on snap deletion

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(snapshot): idempotent snapshot delete

On a retry the subvolume may already have been deleted.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(smoke/snapshot): add snapshot deletion of unstaged volume

Adds a test case where a volume is unstaged when its snapshot is deleted.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented on rawfile/filesystem/btrfs.py:
```diff
@@ -103,7 +103,9 @@ def delete_snapshot(self, name: str):
                         default_subvol_path=$(btrfs subvolume list -o "{tmp_dir}" | awk -v id=$default_subvol_id '$2 == id {{print $NF}}')
                         default_subvol="{tmp_dir}/$default_subvol_path"
-                        btrfs subvolume delete "{snapshot_subvol}"
+                        if [ -f "{snapshot_subvol}" ]; then
```
@mhkarimi1383 the vars created above don't seem to be "used", is it for debugging?

mhkarimi1383 replied:
I think I have forgotten to check them out after converting ops to use commands

tiagolobocastro replied:
Also I wonder if we should not use temporary directory, and just place one in well known location, ex: data folder
Otherwise if let's say the prod crashes during this operation, it'd leak a mount making it very difficult to get rid of.

mhkarimi1383 replied:
I think for these stuff we should do some race tests for example

mhkarimi1383 commented on rawfile/bd2fs.py:
```diff
@@ -251,22 +251,18 @@ def ControllerExpandVolume(self, request, context):
-            snapshot_id = f"{request.source_volume_id}/{request.name}"
-        finally:
-            if fs:
-                fs.unmount(clear_mountpoint=True)
```
Why we do not unmount anymore?

tiagolobocastro replied:
there's nothing to unmount here, the mount which was being caught was from the volume or none if the volume is not staged in the node

mhkarimi1383 replied:
Nice 👍

tiagolobocastro committed:
fix(init): ensure img-file has the correct size

Fixes idempotency of volume creation in case file was created but
not truncated to the size.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(snapshot): cleanup loop device

Add the locks to the snapshot api, which allows us to check if the
source volume is mounted.
If not, then we can cleanup the loop disk device.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
ci(bdd): set correct location of pytest.log

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
ci: erase /mnt

This location contains files from other runs, sometimes leaving us no
free space for the tests.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
@mergifyio queue
